### PR TITLE
Resolves #45 Added notification handler in Rewards Panel

### LINKED ISSUE #45: Show notifications on the main Rewards panel when available
When the user has notifications to view, they should be displayed on `WalletViewController` until dismissed manually (or possibly time based?) 

These notifications may have actions (like #44 for claiming a grant) which can be integrated in different tickets

~Blocked by #41~

### PR DESCRIPTION
**Pull request by danishjafri88**
Implementing #45. 
Borrowing logic from brave-core.
Ref: 
src/brave/components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx
src/brave/components/brave_rewards/resources/extension/brave_rewards/_locales/en_US/messages.json
src/brave/node_modules/brave-ui/src/features/rewards/walletWrapper/index.tsx

### CODE REVIEW

**Comment by danishjafri88** [comment]
Next steps:

- [ ] 1. Verify localisations 
- [ ] 2. Verify Icons and style
- [ ] 3. verify brave-core logic
- [ ] 4. Test with mock data.
- [ ] 5. Unit tests
- [ ] 6. Divide code into separate notification handler?

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        addActionNotification(notification: notification)
+      default:
+        return
+      }
+    } else {
+      delegate?.hideNotifications()
+    }
+  }
+  
+  private func addActionNotification(notification: RewardsNotification) {
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: `add(actionNotification:)` or `addActionNotification(_:)`

File: BraveRewardsUI/Wallet/Notifications/WalletActionNotificationView.swift [hunk 2]
```diff
@@ -10,25 +10,20 @@ struct WalletActionNotification {
...
+    static let grant = Category(
       icon: UIImage(frameworkResourceNamed: "icn-grant"),
       title: Strings.NotificationTokenGrantTitle,
       action: Strings.CLAIM.uppercased()
     )
-    static let recurringTip = Category(
-      icon: UIImage(frameworkResourceNamed: "icn-contribute"),
-      title: Strings.NotificationRecurringTipTitle,
-      action: Strings.OK.uppercased()
+    static let adsLaunch = Category(
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/WalletActionNotificationView.swift]: I dont think we'll need this notification since the plan is to launch with ads

File: BraveRewardsUI/Wallet/Notifications/WalletNotificationView.swift [hunk 3]
```diff
@@ -4,16 +4,63 @@
 
 import UIKit
 
+struct WalletNotification {
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/WalletNotificationView.swift]: If you're going to rename the other from `WalletMessageNotification` to `WalletActionNotification`, why not make this name more specific?

Also why do we now have `alerts`, `actions` and whatever this is? (Message?)

File: BraveRewardsUI/Wallet/Notifications/WalletNotificationView.swift [hunk 4]
```diff
@@ -23,21 +70,72 @@ class WalletNotificationView: UIView {
...
   }
   
   @available(*, unavailable)
   required init(coder: NSCoder) {
     fatalError()
   }
+  
+  private func bodyAttributedString() -> NSAttributedString {
+    guard let notification = notification else {
+      return NSAttributedString()
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/WalletNotificationView.swift]: add `assertionFailure` here

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+      body = Strings.NotificationInsufficientFunds
+      category = .insufficientFunds
+    case .pendingNotEnoughFunds:
+      body = Strings.NotificationPendingNotEnoughFunds
+      category = .insufficientFunds
+    case .adsLaunch:
+      body = Strings.NotificationBraveAdsLaunchMsg
+      category = .adsLaunch
+    default:
+      loadNotification()
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Not sure what this is here for, isn't this undefined behaviour? Shouldn't we just `assertionFaliure`?

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        category: category,
+        body: body,
+        date: date))
+    notificationView.closeButton.addTarget(self, action: #selector(tappedNotificationClose), for: .touchUpInside)
+    notificationView.actionButton.addTarget(self, action: #selector(tappedNotificationAction), for: .touchUpInside)
+    delegate?.show(notificationView: notificationView)
+  }
+  
+  @objc private func tappedNotificationClose() {
+    loadNotification()
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Honestly the fact that we're just calling `loadNotification` to clear the current one is kinda weird to me. Can we just split these out or something?

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+  }
+  
+  @objc private func tappedNotificationAction() {
+    if let notification = currrentNotification {
+      loadNotification()
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Should we be loading the next notification/clearing the current one right away?

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      delegate?.handleAction(notification: notification)
+    }
+  }
+  
+  private func addMessageNotification(notification: RewardsNotification) {
+    var category: WalletNotification.Category = .contribute
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Why's the default value set here as contribute? If there's a reason for it, should leave documentation/comments to explain reasoning

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+      showAlertNotification(type: type, message: body)
+      return
+    }
+    
+    let date = Date(timeIntervalSince1970: notification.dateAdded)
+    let notificationView = WalletNotificationView(
+      notification: WalletNotification(
+        category: category,
+        body: body,
+        date: date))
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: nit: Put trailing parenthesis on their own line

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+  @objc private func notificationAdded(_ notification: Notification) {
+    // TODO: Filter notification?
+    if currrentNotification == nil {
+      loadNotification()
+    }
+  }
+  
+  private func loadNotification() {
+    guard delegate != nil else {
+      return
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: It's not on the class to worry about whether it has a delegate or not. shouldn't return just because its not set. If its vital to be set to have it work properly, it should be required in `init`

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+import Foundation
+import BraveRewards
+
+protocol NotificationHandlerDelegate: AnyObject {
+  func show(notificationView: WalletNotificationView)
+  func handleAction(notification: RewardsNotification)
+  func hideNotifications()
+}
+
+class NotificationHandler {
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Overall, I dont understand this class well…

Its acting as a receiver to turn `RewardsNotification` into one of the `Wallet*Notfication` structs (View Model), in which it then tells the delegate to show notifications (Controller) by creating a view for it (View Controller) and at the same time handle its own notification stack (Model) by clearing away things automatically.

IMO, this class shouldn't be doing all of this, and we should be picking a specific direction. Do we want it to be a "view model store" of sorts, turning `RewardsNotification`s into something that WalletViewController can read and determine what to show?

Whats stopping us of just making this class hold a list of enums that look like this:

```swift
enum WalletNotification
  case alert(AlertNotification)
  case action(ActionNotification)
  case message(MessageNotification)
}
```

On init it just looks at all ledger notifications, converts them, and adds them to its list. At the same time it subscribes to BATBraveLedgerNotificationAdded. Then you can just add simple ledger wrapper methods here that control the list (add, clear, etc.) 

And `WalletViewController` just reads that list in `viewDidLoad` and shows the first one? Then when the user clicks on something we can call `notificationStore.clear(walletNotification)` which will then call the appropriate `BATBraveLedger` method

The only thing that the notification store would still have to do is bind the `WalletViewController` to when the list updates through a delegate method.

I also don't think we should be not subscribing to changes between the point of `init` and `viewDidLoad`. That introduces a small race condition between those two points where we initially read out the notifications in the ledger and not receiving a new notification. This would be fixed by subscribing immediately on init and not forcing the `WalletViewController` to show some view when the list changes even if the view hasn't loaded yet. It'd do it itself in viewDidLoad by just reading out the notification models.

Let me know if we need further clarification around this or if I'm missing something that breaks this thought

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+  deinit {
+    stop()
+  }
+  
+  weak var currrentNotification: RewardsNotification?
+  
+  func start() {
+    // Add observer
+    stop()
+    NotificationCenter.default.addObserver(self, selector: #selector(notificationAdded(_:)), name: .init(rawValue: "BATBraveLedgerNotificationAdded"), object: nil)
```
[the remark this answers is not in the document]
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Did not see it :)

File: BraveRewardsUI/Wallet/Notifications/WalletNotificationView.swift (hunk 3 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/WalletNotificationView.swift]: We have three different kinds of Notifications. A normal message type, another with action and an alert type. Each has its own style and spacing.
I have made WalletNotificationView as the base view by bringing the common elements into it. Basically code inside WalletNotificationView.commonSetup() is not repeated.

Creating another subclass for Message type

File: BraveRewardsUI/Wallet/Notifications/WalletNotificationView.swift (hunk 4 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/WalletNotificationView.swift]: No more required after creating a separate class for message type

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift (hunk 6 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Clearing a notification is always followed by loading next notification. This is coupled so much that clearing a notification in itself does not have a useful case. Also the fact if clear and then set a notification, the animation is not linear but abrupt. (cases inside `setNotificationView`)

Refactoring `loadNotification` to `loadNextNotification` and creating a separate method for clearing notification and replacing it for the clear notification code in `loadNotification`

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift (hunk 8 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Created this method for contribute first and added cases later as I found them, so forgot to remove it. Updated

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift (hunk 10 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Removed, delegate must be set in `start()` for the handler to return callbacks.

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift (hunk 11 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Yeah so its basically a helper class that generates views for notifications and tells that to controller. No stack is maintained in this class, as doing so would require implementing other observers like `delete` etc from braveLedger. 

I created this class to keep WalletViewController decoupled from notification logic and also keep the number of lines in check. But I think the helper class can be further separated into more solid structures. 

I have pushed some refactors. With these changes it gets more clearer.
We have three options from it:
1. Keep as is
2. Let RewardsNotificationService forward notification and not notificationView to WalletViewController and let WalletViewController manage rest.
3. Shift all the code in RewardsNotificationService into WalletViewController.

My reason was that WalletViewController does not depend on Notifications at all. In fact the `notificationView` & `setNotificationView` in `WalletView` are also redundant(Can be simple 
 UIView). So making this class gives us the option to add/remove this feature like an addon.

I am still looking for a better way to do this, let me know what you think is better.

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift (hunk 7 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Broke it down into a callback from receiver before loading next

File: BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift (hunk 11 above)
Comment (kylehickinson) [reply BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: > I created this class to keep WalletViewController decoupled from notification logic

This is why I'm partially confused though because most of the logic is already decoupled because it's in `BATBraveLedger`. It has a stack, of which is managed, and you can call clear on, so all you're doing is abstracting flow logic between `RewardsNotification` and a few different view's per type of `RewardsNotification` from `WalletViewController` into a class which is only used by `WalletViewController` and is unlikely to be used anywhere else since its view specific.

And if the conversion of models to view models is what you want to pull away from WalletViewController thats fine, but the rest of the flow doesn't need to be in there (load next, close button, etc). It could just be in `WalletViewController`

If line count is a big deal you could create a new swift file that just extends `WalletViewController` and call it `WalletViewController+Notifications` or something

> My reason was that WalletViewController does not depend on Notifications at all. In fact the notificationView & setNotificationView in WalletView are also redundant(Can be simple UIView)

This doesn't have to be set in stone though, since again, these notifications are shown nowhere else.  You could just make a `UIViewController` subclass that replaces the NotificationView stuff in WalletViewController.  Then you are essentially only adding logic in `WalletViewController` on whether or not to show said controller based on non-empty notifications in the ledger and maybe getting a _did add notification_ notification which would prompt a check to notifications.

Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/NotificationHandler.swift]: Cool, I moved the code into WalletViewController.

File: BraveRewardsUI/Wallet/Notifications/RewardsNotificationViewBuilder.swift [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        date: date))
+  }
+  
+  private static func get(messageNotification: RewardsNotification) -> WalletNotificationView? {
+    var category: WalletMessageNotification.Category
+    let body: String
+    var alertType: WalletAlertNotification.Category?
+    switch messageNotification.kind {
+    case .autoContribute:
+      if let result = messageNotification.userInfo["result"] as? Int, let amount = messageNotification.userInfo["amount"] as? String {
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/RewardsNotificationViewBuilder.swift]: Use `BATResult` enum switch instead of magic numbers below by creating a `BATResult` from this line

File: BraveRewardsUI/Wallet/WalletViewController.swift [hunk 14]
```diff
@@ -14,6 +14,7 @@ protocol WalletContentView: AnyObject {
 class WalletViewController: UIViewController, RewardsSummaryProtocol {
   
   let state: RewardsState
+  weak var currentNotification: RewardsNotification?
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/WalletViewController.swift]: Why is this `weak`?

File: BraveRewardsUI/Wallet/WalletViewController.swift [hunk 15]
```diff
@@ -364,3 +343,81 @@ class WalletViewController: UIViewController, RewardsSummaryProtocol {
     }
   }
 }
+
+extension WalletViewController {
+  
+  func startNotificationObserver() {
+    // Stopping as a precaution
+    stopNotificationObserver()
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/WalletViewController.swift]: Not necessary if your worry is that you'll a second observer, and since you overwrite `currentNotification` in `loadNextNotification` that's not an issue either

File: BraveRewardsUI/Wallet/WalletViewController.swift [hunk 16]
```diff
@@ -364,3 +343,81 @@ class WalletViewController: UIViewController, RewardsSummaryProtocol {
...
+    // Add observer
+    NotificationCenter.default.addObserver(self, selector: #selector(notificationAdded(_:)), name: NSNotification.Name.BATBraveLedgerNotificationAdded, object: nil)
+    loadNextNotification()
+  }
+  
+  func stopNotificationObserver() {
+    NotificationCenter.default.removeObserver(self, name: NSNotification.Name.BATBraveLedgerNotificationAdded, object: nil)
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/WalletViewController.swift]: Is it possible to shorten this to `.BATBraveLedgerNotificationAdded`?

Comment (kylehickinson) [reply BraveRewardsUI/Wallet/WalletViewController.swift]: Also remind me to add a `NS_SWIFT_NAME` to this in core :P

File: BraveRewardsUI/Wallet/Notifications/RewardsNotificationViewBuilder.swift [hunk 17]
```diff
@@ -0,0 +1,111 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import BraveRewards
+
+struct RewardsNotificationViewBuilder {
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/RewardsNotificationViewBuilder.swift]: This is much better! 

Would it be possible to get some unit tests for the `RewardsNotification` → `WalletActionNotification`/`WalletAlertNotification`/`WalletMessageNotification` logic?

File: BraveRewardsUI/Wallet/WalletViewController.swift [hunk 18]
```diff
@@ -97,6 +79,7 @@ class WalletViewController: UIViewController, RewardsSummaryProtocol {
       navigationController?.setNavigationBarHidden(true, animated: animated)
     }
     reloadUIState()
+    startNotificationObserver()
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/WalletViewController.swift]: Is there a reason this can't be done in `init`?

File: BraveRewardsUI/Wallet/WalletViewController.swift [hunk 19]
```diff
@@ -105,6 +88,11 @@ class WalletViewController: UIViewController, RewardsSummaryProtocol {
     if presentedViewController == nil {
       navigationController?.setNavigationBarHidden(false, animated: animated)
     }
+    stopNotificationObserver()
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/WalletViewController.swift]: Also is there a reason to not update the view anyways even if its not on screen? If a user pops back from settings it would just be there?

File: BraveRewardsUI/Wallet/WalletViewController.swift [hunk 20]
```diff
@@ -364,3 +343,81 @@ class WalletViewController: UIViewController, RewardsSummaryProtocol {
...
+    loadNextNotification()
+  }
+  
+  @objc private func tappedNotificationAction() {
+    if let notification = currentNotification {
+      switch notification.kind {
+      case .grant, .grantAds:
+        tappedGrantsButton()
+      case .pendingNotEnoughFunds, .insufficientFunds:
+        tappedAddFunds()
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/WalletViewController.swift]: We actually _cant_ have the button here for add funds 😞

Since wallet management isn't in scope for this release despite the view already made and working, we'll have to make these message/alert notifications I guess?

File: BraveRewardsUI/Wallet/WalletViewController.swift (hunk 19 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/WalletViewController.swift]: Yeah, since we dont observe for deletedNotification, if any notificationView that may require an action is added and the user performs the action that deletes this notification from ledger will not hide the notificationView. The user has to manually close this notification

File: BraveRewardsUI/Wallet/WalletViewController.swift (hunk 14 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/WalletViewController.swift]: Since the real owner is ledger, and the ledger may remove it from its stack at any time, I found no real reason to keep it strong. If ledger deletes it then we dont have to `deleteNotification` before loading next

File: BraveRewardsUI/Wallet/WalletViewController.swift (hunk 18 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/WalletViewController.swift]: Added in viewdidLoad

File: BraveRewardsTests/RewardsNotificationViewBuilderTests.swift [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+import BraveRewards
+import XCTest
+@testable import BraveRewardsUI
+
+class RewardsNotificationViewBuilderTests: XCTestCase {
+  
+  
+  func testBuilder() {
+    // Notifications not handled:
+    XCTAssertNil(RewardsNotificationViewBuilder.get(notification: notification(.adsLaunch)))
```
Comment (kylehickinson) [inline BraveRewardsTests/RewardsNotificationViewBuilderTests.swift]: Break out these tests into multiple tests (i.e. `testUnhandledNotifications`, `testActionNotifications`, `testMessageNotifications`, `testAlertNotifications`)

File: BraveRewardsUI/Wallet/Notifications/RewardsNotificationViewBuilder.swift [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+      notification: WalletActionNotification(
+        category: category,
+        body: body,
+        date: date))
+  }
+  
+  private static func get(messageNotification: RewardsNotification) -> WalletNotificationView? {
+    var category: WalletMessageNotification.Category!
+    let body: String
+    var alertType: WalletAlertNotification.Category?
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/Notifications/RewardsNotificationViewBuilder.swift]: You've split these out into multiple types and yet in order to get an alert notification you have to call `get(messageNotification: RewardsNotification)`? Isn't there a better way to set this up so there's only 1 type. Is the message + alert notification the same other than the image?

File: BraveRewardsUI/Wallet/WalletViewController.swift [hunk 23]
```diff
@@ -427,3 +405,71 @@ class WalletViewController: UIViewController, RewardsSummaryProtocol {
...
+    if let currentNotification = currentNotification {
+      state.ledger.clearNotification(currentNotification)
+      self.currentNotification = nil
+    }
+  }
+  
+  private func loadNextNotification() {
+    //.failedContribution, .adsLaunch, .invalid & .backupWallet are not supported right now
+    // TODO: Verify that notifications are ordered & they are displayed in FIFO.
+    let ignoredRewardNotificationTypes: [RewardsNotification.Kind] = [.failedContribution,
```
Comment (kylehickinson) [inline BraveRewardsUI/Wallet/WalletViewController.swift]: We have this list of ignored notifications in 2 places now, is it possible to consolidate it somehow?

Comment (danishjafri88) [reply BraveRewardsUI/Wallet/WalletViewController.swift]: I can only see this. Can you mark the other place?

Comment (kylehickinson) [reply BraveRewardsUI/Wallet/WalletViewController.swift]: https://github.com/brave/brave-rewards-ios/pull/123/files#diff-1ed66972ddc2dd1522f79ac2731e9a61R17

Implicitly here

File: BraveRewardsUI/Wallet/Notifications/RewardsNotificationViewBuilder.swift (hunk 22 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/Notifications/RewardsNotificationViewBuilder.swift]: Auto-Contribute notifications can either have a message notification or an alert notification, depending on `BATResult`. Message notification and alert notification have different UI.

File: BraveRewardsExample/ViewController.swift [hunk 24]
```diff
@@ -106,6 +130,12 @@ class ViewController: UIViewController {
     
     setupRewards()
     braveRewardsPanelButton.setImage(RewardsPanelController.batLogoImage, for: .normal)
+//    DispatchQueue.main.asyncAfter(deadline: .now() + 15) {
```
Comment (kylehickinson) [inline BraveRewardsExample/ViewController.swift]: Remove?

File: BraveRewardsTests/RewardsNotificationViewBuilderTests.swift [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+import Foundation
+import BraveRewards
+import XCTest
+@testable import BraveRewardsUI
+
+class RewardsNotificationViewBuilderTests: XCTestCase {
+  
+  
+  func testBuilder() {
```
Comment (kylehickinson) [inline BraveRewardsTests/RewardsNotificationViewBuilderTests.swift]: I'm not sure why you have this?

File: BraveRewardsTests/RewardsNotificationViewBuilderTests.swift [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+   
+    testUnHandledNotifications()
+    // Wanted cases
+    testAlertNotification()
+    testMessageNotification()
+    testActionNotifiication()
+  }
+  
+  func testUnHandledNotifications() {
```
Comment (kylehickinson) [inline BraveRewardsTests/RewardsNotificationViewBuilderTests.swift]: `UnHandled` → `Unhandled`

File: BraveRewardsTests/RewardsNotificationViewBuilderTests.swift [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+    // Notifications not handled:
+    XCTAssertNil(RewardsNotificationViewBuilder.get(notification: notification(.adsLaunch)))
+    
+    XCTAssertNil(RewardsNotificationViewBuilder.get(notification: notification(.backupWallet)))
+
```
Comment (kylehickinson) [inline BraveRewardsTests/RewardsNotificationViewBuilderTests.swift]: Remove whitespace

File: BraveRewardsExample/ViewController.swift (hunk 24 above)
Comment (danishjafri88) [reply BraveRewardsExample/ViewController.swift]: Ah.. was testing postnotification forgot to remove

File: BraveRewardsUI/Wallet/WalletViewController.swift (hunk 23 above)
Comment (danishjafri88) [reply BraveRewardsUI/Wallet/WalletViewController.swift]: well that list is used to filter unwanted builder calls, and first check to stop handling certain notifications. Builder can still have the code to create view for unwanted notifications. Removed filter for now

**Comment by danishjafri88** [comment]
> Incorrect UI design implementation:
> 
> Design	Implementation
> image	image
> image	image
> Not sure if you talked to design about these changes?

Nope I actually took que from desktop. They dont have a button for these types. Rather they had the UI as right now. Will update

**Comment by kylehickinson** [comment]
Would rather just have consistency across platforms so I'm fine with them not matching design if design's fine with it.

**Comment by danishjafri88** [comment]
> Incorrect UI design implementation:
> 
> Design	Implementation
> image	image
> image	image
> Not sure if you talked to design about these changes?

Updated designs after talking with @jamesmudgett, waiting for a final go ahead. 
For reference:

Simulator Screen Shot - iPhone X - 2019-07-17 at 01 22 41
Simulator Screen Shot - iPhone X - 2019-07-17 at 01 22 43
Simulator Screen Shot - iPhone X - 2019-07-17 at 01 22 44
Simulator Screen Shot - iPhone X - 2019-07-17 at 01 22 45
Simulator Screen Shot - iPhone X - 2019-07-17 at 01 22 46
Simulator Screen Shot - iPhone X - 2019-07-17 at 01 22 47
Simulator Screen Shot - iPhone X - 2019-07-17 at 01 25 06

File: BraveRewardsTests/RewardsNotificationViewBuilderTests.swift [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    XCTAssertNil(RewardsNotificationViewBuilder.get(notification: notification(.failedContribution)))
+    XCTAssertNil(RewardsNotificationViewBuilder.get(notification: notification(.invalid)))
+  }
+  
+  func testActionNotifiication() {
+    XCTAssert(RewardsNotificationViewBuilder.get(notification: notification(.grant)) is WalletActionNotificationView)
+  }
+  
+  func testMessageNotification() {
+    XCTAssert(RewardsNotificationViewBuilder.get(notification: notification(.tipsProcessed)) is WalletActionNotificationView)
```
Comment (kylehickinson) [inline BraveRewardsTests/RewardsNotificationViewBuilderTests.swift]: Probably should just move the assertions inside this test into `testActionNotifiication` since there's no "message notification" anymore

Comment (danishjafri88) [reply BraveRewardsTests/RewardsNotificationViewBuilderTests.swift]: Oops, thanks for noticing. Updated now
